// @zeit-ui/react/esm/select/select-option.js
import _extends from "@babel/runtime/helpers/esm/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/esm/objectWithoutProperties";
import _JSXStyle from "styled-jsx/style";
import React, { useMemo } from 'react';
import withDefaults from '../utils/with-defaults';
import useTheme from '../styles/use-theme';
import { useSelectContext } from './select-context';
import useWarning from '../utils/use-warning';
var defaultProps = {
  disabled: false,
  className: '',
  preventAllEvents: false
};

var SelectOption = function SelectOption(_ref) {
  var identValue = _ref.value,
      className = _ref.className,
      children = _ref.children,
      disabled = _ref.disabled,
      preventAllEvents = _ref.preventAllEvents,
      props = _objectWithoutProperties(_ref, ["value", "className", "children", "disabled", "preventAllEvents"]);

  var theme = useTheme();

  var _useSelectContext = useSelectContext(),
      updateValue = _useSelectContext.updateValue,
      value = _useSelectContext.value,
      disableAll = _useSelectContext.disableAll;

  var isDisabled = useMemo(function () {
    return disabled || disableAll;
  }, [disabled, disableAll]);

  if (identValue === undefined) {
    useWarning('The props "value" is required.', 'Select Option');
  }

  var selected = useMemo(function () {
    return value ? identValue === value : false;
  }, [identValue, value]);
  var bgColor = useMemo(function () {
    if (isDisabled) return theme.palette.accents_1;
    return selected ? theme.palette.accents_1 : theme.palette.background;
  }, [selected, isDisabled, theme.palette]);
  var color = useMemo(function () {
    if (isDisabled) return theme.palette.accents_4;
    return selected ? theme.palette.foreground : theme.palette.accents_5;
  }, [selected, isDisabled, theme.palette]);

  var clickHandler = function clickHandler(event) {
    if (preventAllEvents) return;
    event.stopPropagation();
    event.nativeEvent.stopImmediatePropagation();
    event.preventDefault();
    if (isDisabled) return;
    updateValue && updateValue(identValue);
  };

  return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement("div", _extends({
    onClick: clickHandler
  }, props, {
    className: _JSXStyle.dynamic([["1491126873", [theme.layout.gap, theme.layout.gapHalf, bgColor, color, isDisabled ? 'not-allowed' : 'pointer', theme.layout.radius, theme.layout.radius, theme.layout.radius, theme.layout.radius, theme.palette.accents_1]]]) + " " + (props && props.className != null && props.className || "option ".concat(className))
  }), children), /*#__PURE__*/React.createElement(_JSXStyle, {
    id: "1491126873",
    dynamic: [theme.layout.gap, theme.layout.gapHalf, bgColor, color, isDisabled ? 'not-allowed' : 'pointer', theme.layout.radius, theme.layout.radius, theme.layout.radius, theme.layout.radius, theme.palette.accents_1]
  }, ".option.__jsx-style-dynamic-selector{display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;-webkit-box-pack:start;-webkit-justify-content:flex-start;-ms-flex-pack:start;justify-content:flex-start;-webkit-align-items:center;-webkit-box-align:center;-ms-flex-align:center;align-items:center;font-weight:normal;white-space:pre;font-size:0.75rem;height:calc(1.688 * ".concat(theme.layout.gap, ");padding:0 ").concat(theme.layout.gapHalf, ";background-color:").concat(bgColor, ";color:").concat(color, ";-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none;border:0;cursor:").concat(isDisabled ? 'not-allowed' : 'pointer', ";-webkit-transition:background 0.2s ease 0s,border-color 0.2s ease 0s;transition:background 0.2s ease 0s,border-color 0.2s ease 0s;}.option.__jsx-style-dynamic-selector:first-of-type{border-top-left-radius:").concat(theme.layout.radius, ";border-top-right-radius:").concat(theme.layout.radius, ";}.option.__jsx-style-dynamic-selector:last-of-type{border-bottom-left-radius:").concat(theme.layout.radius, ";border-bottom-right-radius:").concat(theme.layout.radius, ";}.option.__jsx-style-dynamic-selector:hover{background-color:").concat(theme.palette.accents_1, ";}")));
};

export default withDefaults(SelectOption, defaultProps);
